// ableton-js/ns/mixer-device.js
import { Namespace } from "./index.js";
import { DeviceParameter } from "./device-parameter.js";
export var PanningMode;
(function (PanningMode) {
    PanningMode[PanningMode["Stereo"] = 0] = "Stereo";
    PanningMode[PanningMode["StereoSplit"] = 1] = "StereoSplit";
})(PanningMode || (PanningMode = {}));
export var CrossfadeAssignment;
(function (CrossfadeAssignment) {
    CrossfadeAssignment[CrossfadeAssignment["A"] = 0] = "A";
    CrossfadeAssignment[CrossfadeAssignment["None"] = 1] = "None";
    CrossfadeAssignment[CrossfadeAssignment["B"] = 2] = "B";
})(CrossfadeAssignment || (CrossfadeAssignment = {}));
export class MixerDevice extends Namespace {
    raw;
    constructor(ableton, raw) {
        super(ableton, "mixer-device", raw.id);
        this.raw = raw;
        this.transformers = {
            crossfader: (v) => new DeviceParameter(ableton, v),
            cue_volume: (v) => new DeviceParameter(ableton, v),
            left_split_stereo: (v) => new DeviceParameter(ableton, v),
            panning: (v) => new DeviceParameter(ableton, v),
            right_split_stereo: (v) => new DeviceParameter(ableton, v),
            sends: (v) => v.map((s) => new DeviceParameter(ableton, s)),
            song_tempo: (v) => new DeviceParameter(ableton, v),
            track_activator: (v) => new DeviceParameter(ableton, v),
            volume: (v) => new DeviceParameter(ableton, v),
        };
    }
}
