// ableton-js/ns/song-view.js
import { Namespace } from "./index.js";
import { Clip } from "./clip.js";
import { ClipSlot } from "./clip-slot.js";
import { DeviceParameter } from "./device-parameter.js";
import { Scene } from "./scene.js";
import { Track } from "./track.js";
export class SongView extends Namespace {
    constructor(ableton) {
        super(ableton, "song-view");
        this.transformers = {
            selected_parameter: (param) => new DeviceParameter(ableton, param),
            selected_track: (track) => new Track(ableton, track),
            selected_scene: (scene) => new Scene(ableton, scene),
            highlighted_clip_slot: (slot) => new ClipSlot(ableton, slot),
            detail_clip: (clip) => new Clip(ableton, clip),
        };
        this.cachedProps = {
            detail_clip: true,
            selected_parameter: true,
            selected_track: true,
            selected_scene: true,
            highlighted_clip_slot: true,
        };
    }
    async selectDevice(device) {
        return this.ableton.sendCommand({
            ns: this.ns,
            name: "select_device",
            args: { device_id: device.raw.id },
        });
    }
}
